src/django_components/component.py

```diff
@@ -1022,101 +1085,173 @@ def _render_impl(
-                component_id=render_id,
-                template_name=template.name,
-                fills=slots_untyped,
-                is_dynamic_component=getattr(self, "_is_dynamic_component", False),
-                # This field will be modified from within `SlotNodes.render()`:
-                # - The `default_slot` will be set to the first slot that has the `default` attribute set.
-                #   If multiple slots have the `default` attribute set, yet have different name, then
-                #   we will raise an error.
-                default_slot=None,
-            )
+            metadata.is_filled = is_filled
 
             with context.update(
                 {
                     # Private context fields
-                    _ROOT_CTX_CONTEXT_KEY: self.outer_context,
-                    _COMPONENT_SLOT_CTX_CONTEXT_KEY: component_slot_ctx,
-                    _REGISTRY_CONTEXT_KEY: self.registry,
+                    _COMPONENT_CONTEXT_KEY: render_id,
                     # NOTE: Public API for variables accessible from within a component's template
                     # See https://github.com/django-components/django-components/issues/280#issuecomment-2081180940
                     "component_vars": ComponentVars(
                         is_filled=is_filled,
                     ),
                 }
             ):
-                self.on_render_before(context, template)
+                # Make a "snapshot" of the context as it was at the time of the render call.
+                #
+                # Previously, we recursively called `Template.render()` as this point, but due to recursion
+                # this was limiting the number of nested components to only about 60 levels deep.
+                #
+                # Now, we make a flat copy, so that the context copy is static and doesn't change even if
+                # we leave the `with context.update` blocks.
+                #
+                # This makes it possible to render nested components with a queue, avoiding recursion limits.
+                context_snapshot = snapshot_context(context)
+
+        # Cleanup
+        context.render_context.pop()
+
+        # Instead of rendering component at the time we come across the `{% component %}` tag
+        # in the template, we defer rendering in order to scalably handle deeply nested components.
+        #
+        # See `_gen_component_renderer()` for more details.
+        deferred_render = self._gen_component_renderer(
+            render_id=render_id,
+            template=template,
+            context=context_snapshot,
+            metadata=metadata,
+            component_path=component_path,
+            css_input_hash=css_input_hash,
+            js_input_hash=js_input_hash,
+            css_scope_id=None,  # TODO - Implement CSS scoping
+        )
+
+        # Remove component from caches
+        def on_component_rendered(component_id: str) -> None:
```

Comment:
These callbacks are passed to `component_post_render` so that this file remains focused on the business logic, while `perfutil/component.py` handles the underlying algo for rendering nested components.